### ISSUE TITLE
Does this code also perform the WGS-84 to GCJ-02 conversion?

### ISSUE THREAD

**Issue by dandv** [maxime/ChinaMapDeviation#1]
Or do I need to use [eviltransform](https://github.com/googollee/eviltransform/issues/6) for that?

**Comment by ghost** [comment]
Yes. But this code seems to be using an interpolation method. At present the accurate algorithm has been leaked, so I would recommend eviltransform and other repos using the algorithm.

**Comment by dandv** [comment]
Ah, so eviltransform does both the coordinates transform and GPS offset/deviation correction?

**Comment by ghost** [comment]
As far as I know, you only need to do coordinates transform. There's no GPS deviation. 

The smartphones Chinese people use also have a bitten Apple on the back of them. The same as your phones. Globalization counts.

**Comment by dandv** [comment]
@fengzee-me,

> There's no GPS deviation.

From http://www.sinosplice.com/life/archives/2012/10/09/unfun-map-games lists:

> the Chinese government is paranoid about the use of GPS, and screws with it. This affects [Google Maps](https://productforums.google.com/forum/?fromgroups=#!topic/maps/zVmRao1Im3w), it affects [camera GPS](http://boingboing.net/2012/05/24/why-your-cameras-gps-wont.html), and it even affects [runners’ watches](http://www.marc.cn/2011/06/chinese-satellite-data-randomly-manipulated.html). It’s been going on for years. Chinese companies with government approval (like TomTom) can get their services/devices working properly, but foreign devices which try to rely on good old-fashioned “satellite positioning” and maps lose out, and have to build in a “shift correction” feature if they want their apps’ GPS positioning to work properly.

That is what I meant by "GPS deviation". I should've called it "GPS shift" perhaps.

http://gis.stackexchange.com/a/59936/37816 mentions a large GPS deviation problem caused by a legal mandate for map API providers to return incorrect data. http://www.sinosplice.com/life/archives/2013/07/16/a-more-complete-ios-solution-to-the-china-gps-offset-problem shows that several people sell massive data sets of tens of thousands of points within China with their corresponding offset.

Are you sure there's no GPS deviation mandated by Chinese law? This deviation could be caused by GPS devices produced in China, to align with maps approved by the Chinese government, or it could be the case that GPS devices produced everywhere return the correct coordinates, but for China maps to align with reality, these coordinates need to be either,
- transformed using [a firmware update](https://web.archive.org/web/20140116215242/http://home.wangjianshuo.com/archives/20081109_all_maps_in_china_are_transformed.htm#comment-31249)
- converted by the China map APIs

Or [are all _other_ map providers forced to show incorrect maps](https://productforums.google.com/forum/?fromgroups=#!topic/maps/zVmRao1Im3w)?

**Comment by dandv** [comment]
Or do [you](http://www.programering.com/a/MzM0MDMwATk.html) mean that GCJ-02 _is_ the same as the "GPS shift"? I found a [paper](http://wenku.baidu.com/view/9a53e4edd5bbfd0a795673c6.html?re=view) you posted on Baidu.

**Comment by ghost** [comment]
Sorry for coming back late. Yes, that's what I mean. The mandatory use of GCJ-02 (and BD-09) is all what happened here in China.

The Chinese government neither bans production of any GPS device, nor makes GPS / GPS chip manufacturers modify their products to report wrong coordinates. The government only gives order to map data providers (or survey companies with government licenses). Those data providers are forced to publish their geo-data under the GCJ-02 system, which makes locations represented in WGS-84 coordinates look wrong when they're pinned to GCJ-02 based digital maps.

Before the WGS-84 <--> GCJ-02 transform algorithm leaked and made available everywhere on the Internet (eg. this repo), some people tried to get a lot of samples of WGS-GCJ point pairs and sold the dataset to people in need. But this kind of dataset is no longer useful because we now have the algorithm itself.

Are all these in the Chinese law?
No. They are mostly in some vague administrative orders or industry regulations made by the government. There is a Code of Geological Survey (http://www.gov.cn/ziliao/flfg/2005-08/05/content_20947.htm, Chinese), though. But it only says data obtained through survey is classified by default (Article 29) and companies without a permission cannot do survey (Chapter 5). Nothing about GCJ-02 is made officially public.

Why does the Chinese government do this?
National security, they say. But people developing LBS apps tends to believe that it is because of money. The government charges Chinese companies for the "shift correction" feature. We need to make a reservation with the government and take the source code of our app to the bureau to get the feature compiled with our code.

**Comment by dandv** [comment]
Thank you @fengzee-me. I've created a Wikipedia article that explains the [China GSP shift problem](https://en.wikipedia.org/wiki/Restrictions_on_geographic_data_in_China).

**Comment by ghost** [comment]
Oh thank you too. You're a good researcher. You can find my old blog post (that have been copied to Baidu by someone) that even I have forgotten about. And it's all Chinese except for a GitHub link. ~_~

Your article explains this problem quite well. I'll make a Chinese version on zh.wikipedia.org once I have time.

**Comment by Artoria2e5** [comment]
Just FYI, I am now working on a zh.wikipedia.org translation for the article created by @dandv: https://zh.wikipedia.org/wiki/Draft:中国大陆对测绘数据的限制.

* * *
Update (2017-04-02): Page has passed DYKC at https://zh.wikipedia.org/wiki/中华人民共和国测绘限制

**Comment by Artoria2e5** [comment]
Umm... Sorry for hijacking the thread again. I currently attributed the iteration method to a last traceable source on GitHub ([caijun/geoChina; cst.r](https://github.com/caijun/geoChina/blob/5c6284b/R/cst.R#L101-L107)), both on Wikipedia and in my [toy implementation](https://artoria2e5.github.io/PRCoords/demo). Should I try to cite @fengzee-me's paper to set the record straight? I might want to do a little bit of [Markdown transcription](https://gist.github.com/anonymous/e7c6f67555099180ce1ae8da4ba2c513) if I go ahead…

**Comment by leojin** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
henrygagnier, @fairlighteth, @PleahMaCaka, @0xDanki, @rlhk, @kschrute, @erikod3r, @ukkary, @403access, @apple-juice, @lukenewman, @nikhilsaraf, @0xAritra, @iAkashPattnaik, @marktanrj, @cozfuttu, @daiwanwei, @AliGharehkhani, @samyoungnyc, @pyncz

**Comment by PleahMaCaka** [comment]
GitHub please

**Comment by leojin** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
mhw32, @iceteapi, @luisalrp, @htadashi, @jxom, @BlakePetersen, @tomoima525, @dhruvmalik007, @blobitty, @egorguscha, @AtlasRW, @imrane, @LilNait-S, @corgii-123, @zercha55, @soundsystems, @k-nrd, @greyireland, @burn777, @qiuqfang
